packages/@aws-cdk-containers/ecs-service-extensions/lib/extensions/queue/queue.ts

```diff
@@ -0,0 +1,393 @@
+ */
+export interface ISubscribable {
+  readonly subscriptionQueue?: SubscriptionQueue;
+  /**
+   * All classes implementing this interface must also implement the `subscribe()` method
+   */
+  subscribe(extension: QueueExtension): sqs.IQueue;
+}
+
+/**
+ * The settings for the Queue extension.
+ */
+export interface QueueExtensionProps {
+  /**
+   * The list of subscriptions for this service.
+   *
+   * @default none
+   */
+  readonly subscriptions?: ISubscribable[];
+
+  /**
+   * The user-provided default queue for this service.
+   *
+   * @default If the `eventsQueue` is not provided, a default SQS Queue is created for the service.
+   */
+  readonly eventsQueue?: sqs.IQueue;
+
+  /**
+   * The user-provided queue delay fields to configure auto scaling for the default queue.
+   */
+  readonly scaleOnLatency?: QueueAutoScalingOptions;
+}
+
+/**
+ * The topic-specific settings for creating the queue subscriptions.
+ */
+export interface TopicSubscriptionProps {
+  /**
+   * The SNS Topic to subscribe to.
+   */
+  readonly topic: sns.ITopic;
+
+  /**
+   * The user-provided queue to subscribe to the given topic.
+   *
+   * @deprecated use `topicSubscriptionQueue`
+   */
+  readonly queue?: sqs.IQueue;
+
+  /**
+   * The object representing topic-specific queue and corresponding queue delay fields to configure auto scaling.
+   * If not provided, the default `eventsQueue` will subscribe to the given topic.
+   *
+   * @default none
+   */
+  readonly topicSubscriptionQueue?: SubscriptionQueue;
+}
+
+/**
+ * `SubscriptionQueue` represents the subscription queue object which includes the topic-specific queue and its
+ * corresponding auto scaling fields.
+ */
+interface SubscriptionQueue {
+  /**
+   * The user-provided queue to subscribe to the given topic.
+   */
+  readonly queue: sqs.IQueue;
+
+  /**
+   * The user-provided queue delay fields to configure auto scaling for the topic-specific queue.
+   *
+   * @default none
+   */
+  readonly scaleOnLatency?: QueueAutoScalingOptions;
+}
+
+/**
+ * Options for configuring SQS Queue auto scaling.
+ */
+interface QueueAutoScalingOptions {
+  /**
+   * Average amount of time for processing a single message in the queue.
+   */
+  readonly messageProcessingTime: cdk.Duration;
+
+  /**
+   * Acceptable amount of time a message can sit in the queue.
+   */
+  readonly acceptableLatency: cdk.Duration;
+}
+
+/**
+ * The `TopicSubscription` class represents an SNS Topic resource that can be subscribed to by the service queues.
+ */
+export class TopicSubscription implements ISubscribable {
+  public readonly topic: sns.ITopic;
+
+  /**
+   * @deprecated use `subscriptionQueue`
+   */
+  public readonly queue?: sqs.IQueue;
+
+  public readonly subscriptionQueue?: SubscriptionQueue;
+
+  constructor(props: TopicSubscriptionProps) {
+    this.topic = props.topic;
+    this.subscriptionQueue = props.topicSubscriptionQueue;
+    this.queue = props.queue ?? props.topicSubscriptionQueue?.queue;
```

Comment:
We should include some validation here so that the customer cannot provide both `queue` and `topicSubscriptionQueue`, rather than having `queue` silently supersede `topicSubscriptionQueue`. If a customer adds the new property, it should be reasonable to ask them to remove the deprecated one.

Reply:
Actually the validation should probably ensure that exactly one of the two is provided. Because this doesn't work if both are undefined, right?

Reply:
I agree, we need a validation to check that both aren't provided together. But it will still work if both are `undefined` as then the subscription will be added for the `eventsQueue`: https://github.com/aws/aws-cdk/pull/17430/files#diff-d67bf88f4c55f19d47ade26933676e9b29d3bf7a250339f13c09fead63499661R141 .